model-optimizer/extensions/middle/ConcatOptimization.py

```diff
@@ -112,8 +108,19 @@ def find_and_replace_pattern(self, graph: Graph):
                     continue
                 shape = in_port.data.get_shape()
                 assert shape is not None
-                if np.array_equal(shape, [0]):
+                if 0 in shape:
                     in_port.disconnect()
 
             connected_input_ports = [in_port for in_port in concat.in_ports().values() if not in_port.disconnected()]
             assert len(connected_input_ports), 'Concat {} does nothing'.format(concat.soft_get('name', concat.id))
+
+            max_port_id = max([port.idx for port in connected_input_ports])
+            if max_port_id != len(connected_input_ports) + 1:
```

Comment:
Is this `if True:` statement? 

Lets say my_concat have connected in ports with indices `[0, 1, 2, 3]`, than `max_port_id == 3`  and `(len(connected_input_ports) + 1)==5`. Algorithm will try to reconnect ports, but everything is fine there.

Maybe we could be more straight forward like this:
`[port.idx for port in connected_input_ports] != list(range(len(connected_input_ports)))`

Reply:
Updated. Actually we need to check that if number of connected ports has changed or their numbers have changed then we need to re-order inputs. So I have updated the check.